File: solutions/Golang/binary_search_tree_iterator_arraylist.go
package main

import "fmt"

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

/**
 * Binary Search Tree Iterator using pre-computed slice.
 * Stores all in-order elements upfront.
 * Space: O(n), next() O(1), hasNext() O(1)
 */
type BSTIterator struct {
    arr   []int
    index int
}

func Constructor(root *TreeNode) BSTIterator {
    iterator := BSTIterator{arr: []int{}, index: 0}
    iterator.inorder(root)
    return iterator
}

func (this *BSTIterator) inorder(node *TreeNode) {
    /**Pre-compute in-order traversal into array.*/
    if node == nil {
        return
    }

    this.inorder(node.Left)
    this.arr = append(this.arr, node.Val)
    this.inorder(node.Right)
}

/**
 * Return next smallest element.
 * Time: O(1)
 */
func (this *BSTIterator) Next() int {
    val := this.arr[this.index]
    this.index++
    return val
}

/**
 * Check if there are more elements.
 * Time: O(1)
 */
func (this *BSTIterator) HasNext() bool {
    return this.index < len(this.arr)
}

func main() {
    fmt.Println("Example 1: BST Iterator using ArrayList")
}

File: solutions/Rust/binary_search_tree_iterator_arraylist.rs
use std::cell::RefCell;
use std::rc::Rc;

#[derive(Debug)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

/**
 * Binary Search Tree Iterator using pre-computed Vec.
 * Stores all in-order elements upfront.
 * Space: O(n), next() O(1), hasNext() O(1)
 */
pub struct BSTIterator {
    arr: Vec<i32>,
    index: usize,
}

impl BSTIterator {
    pub fn new(root: Option<Rc<RefCell<TreeNode>>>) -> Self {
        let mut arr = Vec::new();
        BSTIterator::inorder(root, &mut arr);
        BSTIterator { arr, index: 0 }
    }

    fn inorder(node: Option<Rc<RefCell<TreeNode>>>, arr: &mut Vec<i32>) {
        /**Pre-compute in-order traversal into array.*/
        if let Some(n) = node {
            let node_ref = n.borrow();
            let left = node_ref.left.clone();
            let val = node_ref.val;
            let right = node_ref.right.clone();
            drop(node_ref);

            Self::inorder(left, arr);
            arr.push(val);
            Self::inorder(right, arr);
        }
    }

    /**
     * Return next smallest element.
     * Time: O(1)
     */
    pub fn next(&mut self) -> i32 {
        let val = self.arr[self.index];
        self.index += 1;
        val
    }

    /**
     * Check if there are more elements.
     * Time: O(1)
     */
    pub fn has_next(&self) -> bool {
        self.index < self.arr.len()
    }
}

fn main() {
    println!("Example 1: BST Iterator using ArrayList");
}
